src/openkit-shared/Core/Communication/SessionWrapper.cs

```diff
@@ -0,0 +1,137 @@
+    /// </summary>
+    internal class SessionWrapper
+    {
+        /// <summary>
+        /// The maximum number of "new session requests" to send per session.
+        /// </summary>
+        private const int MaxNewSessionRequests = 4;
+
+        private volatile bool beaconConfigurationSet = false;
+        private volatile bool sessionFinished = false;
+
+        /// <summary>
+        /// Constructor taking the wrapped <see cref="Session"/>.
+        /// </summary>
+        /// <param name="session">The wrapped session.</param>
+        internal SessionWrapper(Session session)
+        {
+            Session = session;
+            NumNewSessionRequestsLeft = MaxNewSessionRequests;
+        }
+
+        /// <summary>
+        /// Get the wrapped session.
+        /// </summary>
+        internal Session Session { get; }
+
+        /// <summary>
+        /// Get a boolean flag indicating whether <see cref="UpdateBeaconConfiguration(BeaconConfiguration)"/>
+        /// has been called before.
+        /// </summary>
+        internal bool IsBeaconConfigurationSet => beaconConfigurationSet;
+
+        /// <summary>
+        /// Get Session's BeaconConfiguration.
+        /// </summary>
+        internal BeaconConfiguration BeaconConfiguration => Session.BeaconConfiguration;
+
+        /// <summary>
+        /// Get a boolean flag indicating whether this session is finished or not.
+        /// </summary>
+        internal bool IsSessionFinished => sessionFinished;
+
+        /// <summary>
+        /// Get and set the Number of new session requests left.
+        /// </summary>
+        internal int NumNewSessionRequestsLeft { get; private set; }
+
+        /// <summary>
+        /// Get a flag indicating whether new session request can be sent, or not.
+        /// </summary>
+        internal bool CanSendNewSessionRequest => NumNewSessionRequestsLeft > 0;
+
+        /// <summary>
+        /// Test if this Session is empty or not.
+        /// </summary>
+        internal bool IsEmpty => Session.IsEmpty;
+
+        /// <summary>
+        /// Get a boolean flag indicating if data sending is allowed or not.
+        /// </summary>
+        internal bool IsDataSendingAllowed => IsBeaconConfigurationSet && BeaconConfiguration.Multiplicity > 0;
+
+        /// <summary>
+        /// Update the beacon's configuration.
+        /// </summary>
+        /// <param name="beaconConfiguration">The new beacon configuration.</param>
+        internal void UpdateBeaconConfiguration(BeaconConfiguration beaconConfiguration)
+        {
+            Session.BeaconConfiguration = beaconConfiguration;
+            beaconConfigurationSet = true;
+        }
+
+        /// <summary>
+        /// Finish this session.
+        /// </summary>
+        internal void FinishSession()
+        {
+            sessionFinished = true;
+        }
+
+        /// Will be called each time a new session request was made for a session.

```

Comment:
like the other comments in this file this comment could be enclosed in ///<summary> and /// </summary>